### ISSUE TITLE
[YSQL] Data lost when use range partition index

### ISSUE THREAD

**Issue by iswarezwp** [yugabyte/yugabyte-db#6457]
I'm migrating an application from mysql to yugabyte. After import data dumped from mysql, the application run into an error:
```sql
SELECT COUNT(id) from order_items where PurchaseDate > '2020-11-20';

Query error: Given ybctid is not associated with any row in table.
```

order_items table has about 7000000 records, and has an index on PurchaseDate:
```
"idx_33539_idx_oi_PurchaseDate" lsm ("PurchaseDate" DESC)
```

**Comment by ddorian** [comment]
Hi @iswarezwp 

Can you describe the schema of the table ? 
Do you mean range sharding ? Or table partitioning ?

**Comment by tedyu** [comment]
From the error message, this seems to be the same issue as #6435

**Comment by tedyu** [comment]
@iswarezwp 
Was idx_33539_idx_oi_PurchaseDate created before the data was inserted or after ?

thanks

**Comment by iswarezwp** [comment]
@tedyu it was created before data was inserted.

**Comment by iswarezwp** [comment]
@ddorian the schema:
```sql
amazon=> \d order_items;
                                           Table "amazon.order_items"
       Column       |           Type           | Collation | Nullable |                 Default                 
--------------------+--------------------------+-----------+----------+-----------------------------------------
 id                 | bigint                   |           | not null | nextval('order_items_id_seq'::regclass)
 OrderItemId        | character varying(50)    |           |          |  
 Title              | character varying(400)   |           |          | 
 QuantityOrdered    | bigint                   |           |          | 
 QuantityShipped    | bigint                   |           |          | 
 Price              | double precision         |           |          | 
 PriceCode          | character varying(20)    |           |          | 
 ShippingPrice      | character varying(100)   |           |          | 
 GiftWrapPrice      | character varying(100)   |           |          | 
 ItemTax            | character varying(100)   |           |          | 
 ShippingTax        | character varying(100)   |           |          | 
 GiftWrapTax        | character varying(100)   |           |          | 
 ShippingDiscount   | character varying(100)   |           |          | 
 PromotionDiscount  | character varying(100)   |           |          | 
 PromotionIds       | character varying(500)   |           |          | 
 IsGift             | boolean                  |           |          | 
 CODFee             | character varying(100)   |           |          | 
 CODFeeDiscount     | character varying(100)   |           |          | 
 GiftMessageText    | character varying(500)   |           |          | 
 GiftWrapLevel      | character varying(100)   |           |          | 
 InvoiceData        | character varying(500)   |           |          | 
 ConditionNote      | character varying(200)   |           |          | 
 ConditionId        | character varying(10)    |           |          | 
 ConditionSubtypeId | character varying(50)    |           |          | 
 ShopId             | bigint                   |           |          | 
 PurchaseDate       | timestamp with time zone |           |          | 
 OrderStatus        | character varying(20)    |           |          | 
 BuyerId            | character varying(50)    |           |          | 
 BuyerName          | character varying(50)    |           |          | 
Indexes:
    "order_items_pkey" PRIMARY KEY, lsm (id)
    "idx_33539_idx_oi_OrderItemId" lsm ("OrderItemId" HASH)
    "idx_33539_idx_oi_PurchaseDate" lsm ("PurchaseDate" DESC)
```

**Comment by iswarezwp** [comment]
@tedyu Data dumped with --inserts option will ok. Maybe batch operation is the problem.

**Comment by nocaway** [comment]
@iswarezwp Which yb release / version are you using?  Was it the beta version?

**Comment by ttyusupov** [comment]
Hi @iswarezwp , we have a fix https://github.com/yugabyte/yugabyte-db/commit/36c2576b679154ee5a4c12e77d371def76dd307b for #6435. 

If you are building yugabyteDB from source code you can try to rebuild and check again if that fixes your issue.

**Comment by iswarezwp** [comment]
Thanks @ttyusupov , I will try it and give a feedback. I use yugabyte 2.3, can I merge https://github.com/yugabyte/yugabyte-db/commit/36c2576b679154ee5a4c12e77d371def76dd307b back to 2.3 with no other changes?

**Comment by nocaway** [comment]
Hi @iswarezwp 
Yugabyte 2.3 has a beta version of RANGE partitioning that has several issues which was reported and fixed in https://github.com/yugabyte/yugabyte-db/issues/4770.

Could you please use YugaByte 2.4 ?

Merging https://github.com/yugabyte/yugabyte-db/commit/36c2576b679154ee5a4c12e77d371def76dd307b back to 2.3 will not fix your issue.

Thank you.

**Comment by iswarezwp** [comment]
OK, thank you @nocaway ,  I will have a try.

**Comment by iswarezwp** [comment]
@nocaway The tserver crashed when test on branch 2.5 with this commit  https://github.com/yugabyte/yugabyte-db/commit/36c2576b679154ee5a4c12e77d371def76dd307b applyed:
```
*** SIGSEGV (@0x0) received by PID 17256 (TID 0x7f804252a700) from PID 0; stack trace: ***
    @     0x7f806e838ba0 (unknown)
    @     0x7f8077b32978  rocksdb::FixedSizeFilterBlockBuilder::AddKey()
    @     0x7f8077b32b91  rocksdb::FixedSizeFilterBlockBuilder::Add()
    @     0x7f8077b17979  rocksdb::BlockBasedTableBuilder::Add()
    @     0x7f8077a2ecf6  rocksdb::CompactionJob::ProcessKeyValueCompaction()
    @     0x7f8077a3049e  rocksdb::CompactionJob::Run()
    @     0x7f8077a6c540  rocksdb::DBImpl::BackgroundCompaction()
    @     0x7f8077a6e14e  rocksdb::DBImpl::BackgroundCallCompaction()
    @     0x7f8077a81147  rocksdb::DBImpl::CompactionTask::DoRun()
    @     0x7f8077a74748  rocksdb::DBImpl::ThreadPoolTask::Run()
    @     0x7f8073fe2957  yb::(anonymous namespace)::PriorityThreadPoolWorker::Run()
    @     0x7f807401042e  yb::Thread::SuperviseThread()
    @     0x7f806e830694 start_thread
    @     0x7f806df6d41d __clone
    @                0x0 (unknown)
```

**Comment by ttyusupov** [comment]
@iswarezwp , could you please, share some (ideally minimal) list of steps on how to reproduce this?

**Comment by ttyusupov** [comment]
Also, is this reproducible on master branch?

**Comment by iswarezwp** [comment]
My build steps:
1. git checkout 2.5
2. git cherry-pick 36c2576b679154ee5a4c12e77d371def76dd307b
3. ./yb_release

and test steps:
1. create 1 cluster with 1 node and rf=1
2. create tables 
3. import data dump from postgres ( I first load data from mysql to postgres using pgloader, and ysql_dump the data from postgres)
4. then load the data with: \i data.sql

**Comment by iswarezwp** [comment]
Only tested on branch 2.5.

**Comment by ttyusupov** [comment]
@iswarezwp , Is it possible to share table creation statements and data.sql (obfuscated if necessary) you using?

**Comment by iswarezwp** [comment]
data.sql is abort 10Gb， I dont know witch part case the problem.

**Comment by ttyusupov** [comment]
Can you try to build YB from the master branch and check that as well?

**Comment by ttyusupov** [comment]
I'll try to reproduce it from my side as well on some sample data

**Comment by iswarezwp** [comment]
@ttyusupov OK, I will test the master branch.

**Comment by iswarezwp** [comment]
@ttyusupov I try the testcase in https://github.com/yugabyte/yugabyte-db/issues/6435， it can also reproduce the problem.

**Comment by ttyusupov** [comment]
@iswarezwp does this test with `generate_series` cause a crash or some other issue? Have you tried that on a build from the fresh master branch?

**Comment by iswarezwp** [comment]
it cause a crash with the same stack:
```
*** Aborted at 1607070864 (unix time) try "date -d @1607070864" if you are using GNU date ***
PC: @     0x7fa3c75d6979 rocksdb::FixedSizeFilterBlockBuilder::AddKey()
*** SIGSEGV (@0x0) received by PID 18570 (TID 0x7fa39959f700) from PID 0; stack trace: ***
    @     0x7fa3be2dcba0 (unknown)
    @     0x7fa3c75d6978  rocksdb::FixedSizeFilterBlockBuilder::AddKey()
    @     0x7fa3c75d6b91  rocksdb::FixedSizeFilterBlockBuilder::Add()
    @     0x7fa3c75bb979  rocksdb::BlockBasedTableBuilder::Add()
    @     0x7fa3c74d2cf6  rocksdb::CompactionJob::ProcessKeyValueCompaction()
    @     0x7fa3c74d449e  rocksdb::CompactionJob::Run()
    @     0x7fa3c7510540  rocksdb::DBImpl::BackgroundCompaction()
    @     0x7fa3c751214e  rocksdb::DBImpl::BackgroundCallCompaction()
    @     0x7fa3c7525147  rocksdb::DBImpl::CompactionTask::DoRun()
    @     0x7fa3c7518748  rocksdb::DBImpl::ThreadPoolTask::Run()
    @     0x7fa3c3a86957  yb::(anonymous namespace)::PriorityThreadPoolWorker::Run()
    @     0x7fa3c3ab442e  yb::Thread::SuperviseThread()
    @     0x7fa3be2d4694 start_thread
    @     0x7fa3bda1141d __clone
    @                0x0 (unknown)

```

I am trying to build the master branch, and will give a feedback later.

**Comment by iswarezwp** [comment]
@ttyusupov master branch has the same problem.

**Comment by iswarezwp** [comment]
It seems that `\i data.sql` (which use COPY command to import data) can quickly reach the problem, `generate_series` need to run 5 ~ 10 minutes.

**Comment by ttyusupov** [comment]
Thanks @iswarezwp , I was able to reproduce that as well, looking into it

**Comment by ttyusupov** [comment]
@iswarezwp I've pushed the fix in commit https://github.com/yugabyte/yugabyte-db/commit/7016365d56ee309b64e94e371e8e7f075c9af7a2, you can check the build from the master branch or use this commit.

**Comment by iswarezwp** [comment]
Thanks @ttyusupov , master branch works fine now.

**Comment by ttyusupov** [comment]
Fixed together with #6435 by commits:
https://github.com/yugabyte/yugabyte-db/commit/36c2576b679154ee5a4c12e77d371def76dd307b
https://github.com/yugabyte/yugabyte-db/commit/7016365d56ee309b64e94e371e8e7f075c9af7a2

**Comment by tedyu** [comment]
Should this be resolved ?
